rust/crd/src/resource.rs

```diff
@@ -0,0 +1,277 @@
+use crate::storage;
+use lazy_static::lazy_static;
+use snafu::Snafu;
+use stackable_operator::{
+    commons::resources::{CpuLimits, MemoryLimits, NoRuntimeLimits, Resources},
+    config::merge::Merge,
+    k8s_openapi::{
+        api::core::v1::ResourceRequirements, apimachinery::pkg::api::resource::Quantity,
+    },
+};
+use strum::{EnumDiscriminants, IntoStaticStr};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum RoleResourceEnum {
+    Druid(Resources<storage::DruidStorage, NoRuntimeLimits>),
+    Historical(Resources<storage::HistoricalStorage, NoRuntimeLimits>),
+}
+
+impl RoleResourceEnum {
+    pub fn as_resource_requirements(&self) -> ResourceRequirements {
+        match self {
+            Self::Druid(r) => r.clone().into(),
+            Self::Historical(r) => r.clone().into(),
+        }
+    }
+    pub fn as_memory_limits(&self) -> MemoryLimits<NoRuntimeLimits> {
+        match self {
+            Self::Druid(r) => r.clone().memory,
+            Self::Historical(r) => r.clone().memory,
+        }
+    }
+}
+
+#[derive(Snafu, Debug, EnumDiscriminants, PartialEq, Eq)]
+#[strum_discriminants(derive(IntoStaticStr))]
+#[allow(clippy::enum_variant_names)]
+pub enum Error {
+    #[snafu(display("no resources available for merging"))]
+    NoResourcesToMerge,
+    #[snafu(display("cannot merge storage types of different roles"))]
+    IncompatibleStorageMerging,
+}
+
+/// Merge resources from left to right: first > second > third.
+/// Return a copy of the merged struct.
+pub fn try_merge(
+    first: Option<&mut RoleResourceEnum>,
+    second: Option<&mut RoleResourceEnum>,
+    third: Option<&mut RoleResourceEnum>,
+) -> Result<RoleResourceEnum, Error> {
+    let mut some = [first, second, third]
+        .into_iter()
+        .flatten()
+        .collect::<Vec<&mut RoleResourceEnum>>();
+
+    match some.len() {
```

Comment:
There might be more clever ways but fine for me for 3 entries

Reply:
If you can find a clever way that's also the same amount of lines of code or even less, I'm interested :)

Reply:
I pushed a proposal to the `role-based-configuration-review` branch ;)
Might be something for operator-rs